### ISSUE TITLE
`Emit`/`Json Object to Struct` on every `Tick` produces poor frame rate

### ISSUE THREAD

**Issue by dionysiusmarquis** [getnamo/socketio-client-unreal#162]
I need to send full `skeletal data`, including complete bone transforms etc via socket on every `Tick`. I already had `Socket-IO` set up for another purpose and created a prove of concept with this plugin in `Blueprint` context. As soon as I call any `Emit` on a `SocketIOClientComponent` on every `Tick` with the same or newly created `SIOJsonValue` the frame rate drops. the amount of data and connection speed matters. More data and/or slower connection the more frame drops. I have no noticeable or significant CPU usage. It seems `Emit` does block the `Game Thread`. Is that true/expected?
Is there a way to verify "`Multi-threaded` and non-blocking by default".

**Comment by getnamo** [comment]
Plugin/Unreal version? Do you have a minimal project that replicates this issue so I can see what kind of data throughput and blueprint nodes tank the framerate.

In general emits called from a blueprint get called on a gamethread before having the data passed on to the sio network thread. It's possible the conversion of sio message types which takes place on the game thread may cause performance issues, but I haven't personally seen these cases yet (its possible I haven't pushed enough data at high enough rates).

Note to self: check if https://github.com/getnamo/socketio-client-ue4/blob/master/Source/SocketIOClient/Private/SocketIONative.cpp#L226 can be scheduled on background thread before conversions. Possible choke point at e.g.: https://github.com/getnamo/socketio-client-ue4/blob/master/Source/SocketIOClient/Private/SocketIONative.cpp#L145

**Comment by dionysiusmarquis** [comment]
Unfortunately I'll not find the time to create a simple UE Project at the moment. I switched to  [Simple UDP TCP Socket Server](https://www.unrealengine.com/marketplace/en-US/slug/simple-udp-tcp-socket-server) and [Simple UDP TCP Socket Client
](https://www.unrealengine.com/marketplace/en-US/slug/simple-udp-tcp-socket-client) which utilize the Unreal Engine `FSocket` without any additional dependencies. This fixed the frame drops as it doesn't execute in the `Game Thread` as well. 
I still used `SIOJson` to convert `Struct` to `string` and vise versa. This let me to a more easily replicating and definitive `Game Thread` blocker: Take a multi-nested "big" JSON `string` and convert it via `string` -> `Value From Json String` -> `To Object (JsonValue)` -> `Json Object to Struct` and execute it every Tick. This will completely block the `Game Thread`. The other way around (`Struct` to JSON `string`) is no problem at all. First I thought it could be the Blueprint to C++ Struct "Hack":
https://github.com/getnamo/socketio-client-ue4/blob/d3179b3095196b99f097532e7664bf83d242c027/Source/SIOJson/Public/SIOJLibrary.h#L108-L111
I made several blueprint helper functions to fill the `Struct` explicit with nested `Get Object Field` calls using the known nested JSON `key:value` structure. This will still completely block the `Game Thread`. So maybe it is the Unreal `FJson*` System which is very poor with "big" nested structures? I then switched to [Kantan AutoJSON](https://www.unrealengine.com/marketplace/en-US/slug/kantan-autojson) which uses a slim non Unreal related JSON library, now everything runs smooth.

I was broadcasting the SocketIO `Event` to all clients including the sender so maybe it was `SIOJson`/`FJson*` that blocked the `Game Thread` all along?
